deepset_cloud_sdk/workflows/pipeline_client/pipeline_service.py

```diff
@@ -0,0 +1,257 @@
+"""Pipeline publishing service for deepset Cloud SDK."""
+from contextlib import asynccontextmanager
+from typing import Any, AsyncGenerator, Dict, Protocol, runtime_checkable
+
+import structlog
+
+from deepset_cloud_sdk._api.config import DEFAULT_WORKSPACE_NAME, CommonConfig
+from deepset_cloud_sdk._api.deepset_cloud_api import DeepsetCloudAPI
+from deepset_cloud_sdk.workflows.pipeline_client.models import (
+    PipelineInputs,
+    PipelineOutputs,
+    PipelineType,
+    PublishConfig,
+)
+from deepset_cloud_sdk.workflows.user_facing_docs.pipeline_service_docs import (
+    PipelineServiceDocs,
+)
+
+logger = structlog.get_logger(__name__)
+
+
+@runtime_checkable
+class PipelineProtocol(Protocol):
+    """Protocol defining the required methods for a pipeline object."""
+
+    def dumps(self) -> str:
+        """Convert the pipeline to a YAML string.
+
+        :return: YAML string representation of the pipeline
+        """
+        ...
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Convert the pipeline to a dictionary representation.
+
+        :return: Dictionary representation of the pipeline
+        """
+        ...
+
+    def add_component(self, name: str, instance: Any) -> None:
+        """Add a component to the pipeline.
+
+        :param name: Name of the component
+        :param instance: Component instance to add
+        """
+        ...
+
+    def connect(self, sender: str, receiver: str) -> "PipelineProtocol":
+        """Connect two components in the pipeline.
+
+        :param sender: Component name or component.connection_name that sends data
+        :param receiver: Component name or component.connection_name that receives data
+        :return: The pipeline instance for method chaining
+        """
+        ...
+
+
+class PipelineService:
+    """Handles the publishing of Haystack pipelines and indexes to deepset Cloud."""
+
+    def __init__(self, api: DeepsetCloudAPI) -> None:
+        """Initialize the pipeline service.
+
+        :param api: An initialized DeepsetCloudAPI instance
+        """
+        self.api = api
+
+    @classmethod
+    async def factory(cls, config: CommonConfig) -> "PipelineService":
+        """Create a new instance of the pipeline service.
+
+        :param config: CommonConfig object.
+        """
+        async with DeepsetCloudAPI.factory(config) as api:
+            return cls(api)
+
+    def _append_section_to_yaml(self, yaml_str: str, section_name: str, fields: dict) -> str:
+        """Append a section configuration to the YAML string.
+
+        :param yaml_str: Original YAML string
+        :param section_name: Name of the section to append (e.g., 'inputs' or 'outputs')
+        :param fields: Dictionary of fields to append
+        :return: YAML string with section appended
+        """
+        # Remove empty values
+        fields = {k: v for k, v in fields.items() if v}
```

Comment:
why not using something like ruamal.yaml / yaml here?

Reply:
refactored to use ruamal.yaml. Will come in handy when adding more transformations.